internal/enricher/geolocation.go

```diff
@@ -0,0 +1,235 @@
+package enricher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/filemanager"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	"github.com/rudderlabs/rudder-server/services/geolocation"
+	"github.com/rudderlabs/rudder-server/utils/types"
+)
+
+const (
+	ERR_INVALID_IP    = "invalid_ip"
+	ERR_EMPTY_IP      = "empty_ip"
+	ERR_LOCATE_FAILED = "locate_failed"
+)
+
+type Geolocation struct {
+	City     string `json:"city"`
+	Country  string `json:"country"`
+	Region   string `json:"region"`
+	Postal   string `json:"postal"`
+	Location string `json:"location"`
+	Timezone string `json:"timezone"`
+}
+
+type geoEnricher struct {
+	fetcher geolocation.GeoFetcher
+	logger  logger.Logger
+	stats   stats.Stats
+}
+
+func NewGeoEnricher(conf *config.Config, log logger.Logger, statClient stats.Stats) (PipelineEnricher, error) {
+	log.Infof("Setting up new geo enricher")
+
+	dbPath, err := downloadMaxmindDB(context.Background(), conf, log)
+	if err != nil {
+		return nil, fmt.Errorf("downloading instance of maxmind db: %w", err)
+	}
+
+	fetcher, err := geolocation.NewMaxmindDBReader(dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("creating new instance of maxmind's geolocation db reader: %w", err)
+	}
+
+	return &geoEnricher{
+		fetcher: fetcher,
+		stats:   statClient,
+		logger:  log.Child("geolocation"),
+	}, nil
+}
+
+// Enrich function runs on a request of GatewayBatchRequest which contains
+// multiple singular events from a source. The enrich function augments the
+// geolocation information per event based on IP address.
+func (e *geoEnricher) Enrich(source *backendconfig.SourceT, request *types.GatewayBatchRequest) error {
+	if !source.GeoEnrichment.Enabled {
+		return nil
+	}
+	e.logger.Debugw("received a call to enrich gateway events for source", "sourceID", source.ID)
+
+	var (
+		rawGeo *geolocation.GeoInfo
+		err    error
+	)
+
+	defer func() {
+		errType := ""
+
+		if request.RequestIP == "" {
+			errType = ERR_EMPTY_IP
+		}
+
+		if err != nil {
+			errType = ERR_LOCATE_FAILED
+
+			if errors.Is(err, geolocation.ErrInvalidIP) {
+				errType = ERR_INVALID_IP
+			}
+		}
+
+		e.stats.NewTaggedStat(
+			"proc_geo_enricher_request",
+			stats.CountType,
+			stats.Tags{
+				"sourceID": source.ID,
+				"error":    errType,
+			}).Increment()
+	}()
+
+	if request.RequestIP == "" {
```

Comment:
According to latest discoveries, the message itself could contain a `context.ip` field which should take precedence over the request's ip.